src/com/facebook/buck/testrunner/DelegateRunnerWithTimeout.java

```diff
@@ -126,9 +129,22 @@ private void runWithBuckManagedTimeout(DelegateRunNotifier wrapper) {
     // the default timeout, we cancel the Runner to protect against the case where the test hangs
     // forever.
     while (true) {
-      if (testsCompleted.get()) {
-        // Normal termination: hooray!
-        return;
+      if (testCompleted.isDone()) {
+        try {
+          testCompleted.join();
+          // Normal termination: hooray!
+          return;
+        } catch (CompletionException completionException) {
+          // Unwrap and re-throw original RuntimeException or Error
+          if (completionException.getCause() instanceof RuntimeException) {
```

Comment:
nit: Use one of these?
https://google.github.io/guava/releases/21.0/api/docs/com/google/common/base/Throwables.html#throwIfInstanceOf-java.lang.Throwable-java.lang.Class-